### ISSUE TITLE
memory leak

### ISSUE THREAD

**Issue by ioir123ju** [pytorch/pytorch#38717]
## 🐛 Bug

 torch.zeros has a memory leak!

## To Reproduce

When I run the torch.zeros, the memory grows all the way up to 64 G

```
# demo
while True:
    bboxs.size(0) = 1 or 2
    self.input_size = (250, 192)
    inps = torch.zeros(bboxs.size(0), 3, *self.input_size)
```

## Environment

PyTorch version: 1.4.0+cu100
Is debug build: No
CUDA used to build PyTorch: 10.0

OS: Ubuntu 18.04.3 LTS
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
CMake version: version 3.12.2

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration: 
GPU 0: Tesla V100-PCIE-16GB
GPU 1: Tesla V100-PCIE-16GB

Nvidia driver version: 410.129
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.0

Versions of relevant libraries:
[pip] numpy==1.18.1
[pip] torch==1.4.0+cu100
[pip] torch2trt==0.0.3
[pip] torchfile==0.1.0
[pip] torchnet==0.0.4
[pip] torchreid==1.0.9
[pip] torchvision==0.5.0+cu100
[conda] _tflow_select             2.3.0                       mkl    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] blas                      1.0                         mkl    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] cudatoolkit               10.0.130                      0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] mkl                       2019.4                      243    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] mkl-service               2.3.0            py36he904b0f_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] mkl_fft                   1.0.14           py36ha843d7b_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] mkl_random                1.0.2            py36hd81dba3_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] numpy                     1.18.1                   pypi_0    pypi
[conda] tensorflow                1.15.0          mkl_py36h4920b83_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] torch                     1.3.1                    pypi_0    pypi
[conda] torch2trt                 0.0.3                    pypi_0    pypi
[conda] torchfile                 0.1.0                    pypi_0    pypi
[conda] torchnet                  0.0.4                    pypi_0    pypi
[conda] torchreid                 1.0.9                     dev_0    <develop>
[conda] torchvision               0.5.0+cu100              pypi_0    pypi

```

 - PyTorch Version (e.g., 1.0):
 - OS (e.g., Linux):
 - How you installed PyTorch (`conda`, `pip`, source):
 - Build command you used (if compiling from source):
 - Python version:
 - CUDA/cuDNN version:
 - GPU models and configuration:
 - Any other relevant information:

**Comment by glaringlee** [comment]
@ioir123ju
Thanks for filing an issue. Can you post your full demo code? So we can reproduce this error on our end. I wrote a similar testing code, but I was not able to reproduce it.

**Comment by glaringlee** [comment]
Mark triage review for further investigation.

**Comment by ioir123ju** [comment]
> @ioir123ju
> Thanks for filing an issue. Can you post your full demo code? So we can reproduce this error on our end. I wrote a similar testing code, but I was not able to reproduce it.

I tested it again and memory grows by more than 2 G in two hours.
full demo code
```
import torch
import random
while True:

    bbox = random.randint(1, 2)
    inps = torch.zeros(bbox, 3, *(256, 192))
```
![image](https://user-images.githubusercontent.com/14951957/82399548-f5017f80-9a87-11ea-8097-8d49a88ea839.png)
2 hours later
![image](https://user-images.githubusercontent.com/14951957/82399517-e2874600-9a87-11ea-8979-8a058afb8241.png)

**Comment by amandeep1991** [comment]
Hi @ioir123ju / @glaringlee ,

I have tried this for version : ''1.5.0+cpu'' on windows and it seems there is no such memory leak happening, the issue was raised for version: "1.4.0+cu100", so I don't think a fix is required.

Thanks,
Aman

**Comment by glaringlee** [comment]
@amandeep1991 
Thanks Aman, you are right, I noticed that, but I will still put this in our triage meeting to make sure this is not an issue in v1.5. I can not reproduce this under v1.5 + gpu as well, but I don't have v100 on hand, can not test with same env. Will post an update here once this ticket is triaged.

**Comment by ioir123ju** [comment]
> @amandeep1991 @ioir123ju
> Thanks Aman, you are right, I noticed that, but I will still put this in our triage meeting to make sure this is not an issue in v1.5. I can not reproduce this under v1.5 + gpu as well, but I don't have v100 on hand, can not test with same env. Will post an update here once this ticket is triaged.

### talk about the conclusion first
I did a control experiment, and find which version is wrong.
`pip install torch==1.4.0+cu100 torchvision==0.5.0+cu100 -f https://download.pytorch.org/whl/torch_stable.html`
Use this command to install pytorch1.4 will have a memory leak.
`conda install pytorch torchvision cudatoolkit=10.0`
Use this command to install pytorch1.4 is ok.
### V100
**Environment 1：**
PyTorch version: 1.4.0+cu100
Is debug build: No
CUDA used to build PyTorch: 10.0
Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration: 
GPU 0: Tesla V100-PCIE-16GB
GPU 1: Tesla V100-PCIE-16GB

Nvidia driver version: 410.129
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.0
[conda] torch                     1.3.1                    pypi_0    pypi
[conda] torchvision               0.4.0                    pypi_0    pypi

**has memory leak **
![微信截图_20200527093148](https://user-images.githubusercontent.com/14951957/82968826-c2640380-a000-11ea-958b-509330e24463.png)

**Environment 2：**
PyTorch version: 1.4.0
Is debug build: No
CUDA used to build PyTorch: 10.0
Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration: 
GPU 0: Tesla V100-PCIE-16GB
GPU 1: Tesla V100-PCIE-16GB
Nvidia driver version: 410.129
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.0
[conda] pytorch                   1.4.0           py3.6_cuda10.0.130_cudnn7.6.3_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/pytorch
[conda] torchvision               0.5.0                py36_cu100    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/pytorch


**no memory leak happening**
![v100_torch1 4](https://user-images.githubusercontent.com/14951957/82968870-d871c400-a000-11ea-9301-823b3a398466.png)

### T4
**Environment 1：**
PyTorch version: 1.4.0
Is debug build: No
CUDA used to build PyTorch: 10.0
Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration: 
GPU 0: Tesla T4
GPU 1: Tesla T4
Nvidia driver version: 410.129
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.0
[pip] numpy==1.18.1
[pip] torch==1.4.0
[pip] torchvision==0.5.0
[conda] pytorch                   1.4.0           py3.6_cuda10.0.130_cudnn7.6.3_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/pytorch
[conda] torchvision               0.5.0                py36_cu100    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/pytorch

**no memory leak happening**
![微信截图_20200527093659](https://user-images.githubusercontent.com/14951957/82968279-854b4180-9fff-11ea-857c-50981ba998eb.png)

**Environment 2：**
PyTorch version: 1.2.0
Is debug build: No
CUDA used to build PyTorch: 10.0.130
Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration: 
GPU 0: Tesla T4
GPU 1: Tesla T4

Nvidia driver version: 410.129
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.0
[pip] numpy==1.17.3
[pip] torch==1.2.0
[pip] torchvision==0.2.1
[conda] pytorch                   1.2.0           cuda100py36h938c94c_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
[conda] torchfile                 0.1.0                    pypi_0    pypi
[conda] torchvision               0.6.0                    pypi_0    pypi

**no memory leak happening**
![微信截图_20200527091736](https://user-images.githubusercontent.com/14951957/82968797-b37d5100-a000-11ea-9243-27d92f42fd28.png)

**Environment 3：**
PyTorch version: 1.4.0+cu100
Is debug build: No
CUDA used to build PyTorch: 10.0
Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration: 
GPU 0: Tesla T4
GPU 1: Tesla T4

Nvidia driver version: 410.129
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.0
[pip] torch==1.4.0+cu100
[pip] torchvision==0.5.0+cu100
[conda] torch                     1.4.0+cu100              pypi_0    pypi
[conda] torchvision               0.5.0+cu100              pypi_0    pypi

**has memory leak**
![T4](https://user-images.githubusercontent.com/14951957/82973335-a7e25800-a009-11ea-812c-59a4f3b1a742.png)

**Comment by ioir123ju** [comment]
> > @amandeep1991 @ioir123ju
> > Thanks Aman, you are right, I noticed that, but I will still put this in our triage meeting to make sure this is not an issue in v1.5. I can not reproduce this under v1.5 + gpu as well, but I don't have v100 on hand, can not test with same env. Will post an update here once this ticket is triaged.
> 
> ### talk about the conclusion first
> I did a control experiment, and find which version is wrong.
> `pip install torch==1.4.0+cu100 torchvision==0.5.0+cu100 -f https://download.pytorch.org/whl/torch_stable.html`
> Use this command to install pytorch1.4 will have a memory leak.
> 
> `conda install pytorch torchvision cudatoolkit=10.0`
> Use this command to install pytorch1.4 is ok.
> 
> ### V100
> **Environment 1：**
> PyTorch version: 1.4.0+cu100
> Is debug build: No
> CUDA used to build PyTorch: 10.0
> Python version: 3.6
> Is CUDA available: Yes
> CUDA runtime version: 10.0.130
> GPU models and configuration:
> GPU 0: Tesla V100-PCIE-16GB
> GPU 1: Tesla V100-PCIE-16GB
> 
> Nvidia driver version: 410.129
> cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.0
> [conda] torch 1.3.1 pypi_0 pypi
> [conda] torchvision 0.4.0 pypi_0 pypi
> 
> **has memory leak**
> ![微信截图_20200527093148](https://user-images.githubusercontent.com/14951957/82968826-c2640380-a000-11ea-958b-509330e24463.png)
> 
> **Environment 2：**
> PyTorch version: 1.4.0
> Is debug build: No
> CUDA used to build PyTorch: 10.0
> Python version: 3.6
> Is CUDA available: Yes
> CUDA runtime version: 10.0.130
> GPU models and configuration:
> GPU 0: Tesla V100-PCIE-16GB
> GPU 1: Tesla V100-PCIE-16GB
> Nvidia driver version: 410.129
> cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.0
> [conda] pytorch 1.4.0 py3.6_cuda10.0.130_cudnn7.6.3_0 https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/pytorch
> [conda] torchvision 0.5.0 py36_cu100 https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/pytorch
> 
> **no memory leak happening**
> ![v100_torch1 4](https://user-images.githubusercontent.com/14951957/82968870-d871c400-a000-11ea-9301-823b3a398466.png)
> 
> ### T4
> **Environment 1：**
> PyTorch version: 1.4.0
> Is debug build: No
> CUDA used to build PyTorch: 10.0
> Python version: 3.6
> Is CUDA available: Yes
> CUDA runtime version: 10.0.130
> GPU models and configuration:
> GPU 0: Tesla T4
> GPU 1: Tesla T4
> Nvidia driver version: 410.129
> cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.0
> [pip] numpy==1.18.1
> [pip] torch==1.4.0
> [pip] torchvision==0.5.0
> [conda] pytorch 1.4.0 py3.6_cuda10.0.130_cudnn7.6.3_0 https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/pytorch
> [conda] torchvision 0.5.0 py36_cu100 https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/pytorch
> 
> **no memory leak happening**
> ![微信截图_20200527093659](https://user-images.githubusercontent.com/14951957/82968279-854b4180-9fff-11ea-857c-50981ba998eb.png)
> 
> **Environment 2：**
> PyTorch version: 1.2.0
> Is debug build: No
> CUDA used to build PyTorch: 10.0.130
> Python version: 3.6
> Is CUDA available: Yes
> CUDA runtime version: 10.0.130
> GPU models and configuration:
> GPU 0: Tesla T4
> GPU 1: Tesla T4
> 
> Nvidia driver version: 410.129
> cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.0
> [pip] numpy==1.17.3
> [pip] torch==1.2.0
> [pip] torchvision==0.2.1
> [conda] pytorch 1.2.0 cuda100py36h938c94c_0 https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main
> [conda] torchfile 0.1.0 pypi_0 pypi
> [conda] torchvision 0.6.0 pypi_0 pypi
> 
> **no memory leak happening**
> ![微信截图_20200527091736](https://user-images.githubusercontent.com/14951957/82968797-b37d5100-a000-11ea-9243-27d92f42fd28.png)
> 
> **Environment 3：**
> PyTorch version: 1.4.0+cu100
> Is debug build: No
> CUDA used to build PyTorch: 10.0
> Python version: 3.6
> Is CUDA available: Yes
> CUDA runtime version: 10.0.130
> GPU models and configuration:
> GPU 0: Tesla T4
> GPU 1: Tesla T4
> 
> Nvidia driver version: 410.129
> cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.0
> [pip] torch==1.4.0+cu100
> [pip] torchvision==0.5.0+cu100
> [conda] torch 1.4.0+cu100 pypi_0 pypi
> [conda] torchvision 0.5.0+cu100 pypi_0 pypi
> 
> **has memory leak**
> ![T4](https://user-images.githubusercontent.com/14951957/82973335-a7e25800-a009-11ea-812c-59a4f3b1a742.png)

**Comment by glaringlee** [comment]
@ioir123ju 
Can you try torch 1.5? I saw all your test is based on torch 1.4. This issue might be already fixed under torch 1.5, since I can not reproduce it under 1.5.

**Comment by ioir123ju** [comment]
@glaringlee
I can't test torch 1.5. Because my CUDA version is 10.0. 
That's all I can do

**Comment by ezyang** [comment]
I think all of the testing here indicates that it is fixed in 1.5. Unfortunately, if your version of CUDA is too old, you'll have to upgrade CUDA, or perhaps try building PyTorch yourself with the older version of CUDA.

**Comment by ezyang** [comment]
BTW, you said:

> I tested it again and memory grows by more than 2 G in two hours.

This implies that it is a very slow leak. That suggests that it's not CUDA memory that we're leaking, but some sort of small Python object. I think we have fixed a few cases where we were leaking these on master.
